distsys/archetypeinterface.go

```diff
@@ -0,0 +1,218 @@
+package distsys
+
+import (
+	"fmt"
+	"github.com/benbjohnson/immutable"
+)
+
+type ArchetypeInterface struct {
+	ctx *MPCalContext
+}
+
+func (iface ArchetypeInterface) Self() TLAValue {
+	return iface.ctx.self
+}
+
+func (iface ArchetypeInterface) ensureCriticalSectionWith(handle ArchetypeResourceHandle) {
+	iface.ctx.dirtyResourceHandles[handle] = true
+}
+
+func (iface ArchetypeInterface) Write(handle ArchetypeResourceHandle, indices []TLAValue, value TLAValue) (err error) {
+	iface.ensureCriticalSectionWith(handle)
+	res := iface.ctx.getResourceByHandle(handle)
+	for _, index := range indices {
+		res, err = res.Index(index)
+		if err != nil {
+			return
+		}
+	}
+	err = res.WriteValue(value)
+	return
+}
+
+func (iface ArchetypeInterface) Read(handle ArchetypeResourceHandle, indices []TLAValue) (value TLAValue, err error) {
+	iface.ensureCriticalSectionWith(handle)
+	res := iface.ctx.getResourceByHandle(handle)
+	for _, index := range indices {
+		res, err = res.Index(index)
+		if err != nil {
+			return
+		}
+	}
+	value, err = res.ReadValue()
+	return
+}
+
+func (iface ArchetypeInterface) Fairness(id string, ceiling int) int {
+	fairnessCounters := iface.ctx.fairnessCounters
+	counter := fairnessCounters[id]
+	var nextCounter int
+	if counter >= ceiling - 1 {
+		nextCounter = 0
+	} else {
+		nextCounter = counter + 1
+	}
+	fairnessCounters[id] = nextCounter
+	return counter
+}
+
+func (iface ArchetypeInterface) GetConstant(name string) func(args... TLAValue) TLAValue {
+	fn, wasFound := iface.ctx.constantDefns[name]
+	if !wasFound {
+		panic(fmt.Errorf("could not find constant definition %s", name))
+	}
+	return fn
+}
+
+func (iface ArchetypeInterface) RequireArchetypeResource(name string) ArchetypeResourceHandle {
+	handle := ArchetypeResourceHandle(name)
+	_ = iface.ctx.getResourceByHandle(handle)
+	return handle
+}
+
+func (iface ArchetypeInterface) RequireArchetypeResourceRef(name string) (ArchetypeResourceHandle, error) {
+	ptr := iface.RequireArchetypeResource(name)
+	ptrVal, err := iface.Read(ptr, nil)
+	if err != nil {
+		return "", err
+	}
+	return iface.RequireArchetypeResource(ptrVal.AsString()), nil
+}
+
+func (iface ArchetypeInterface) EnsureArchetypeResourceLocal(name string, value TLAValue) {
+	_ = iface.ctx.ensureArchetypeResource(name, LocalArchetypeResourceMaker(value))
+}
+
+func (iface ArchetypeInterface) ReadArchetypeResourceLocal(name string) TLAValue {
```

Comment:
For the PreAmble of an archetype, in the rare case where you get:
```
variables x = 1; y = x + 2;
```
Which has to compile the access for `x` in the definition of `y` to something... but PreAmble isn't really a critical section, so we do some encapsulation violation to _just do the access_ without the error handling, because local resources are very simple and kind of a special case.

I think it's used in ~1 other case, but basically that's the issue. Bypassing the critical section mechanism in archetype PreAmble.